### PR TITLE
[alpaka] Support CUDA or ROCm/HIP

### PR DESCRIPTION
**Pull request by fwyzard**
Implement the changes for building the `alpakatest` and `alpaka` applications with support for either of CUDA or HPI/ROCm.

### Host code changes

1. Add Cuda and Hip types

Since I hope to be able to enable both CUDA and HIP/ROCm at some point in the future, I've decided to split already now the relevant Alpaka types.
Alpaka itself does a mixed effort on this:
  - the "accelerator" types are already available as two distinct types
  - some other types are just `using` aliases for the common type
  - some other types are not available at all

I've added these last ones in `src/alpaka/alpaka/alpakaExtra.hpp`, with the intention of moving them into Alpaka itself sooner or later.

2. Replace the use of the UniformCudaHipRt types with the explicit CudaRt types

3. Add similar code paths and definitions for the HipRt types

I've duplicated all CUDA-specific code, that was using either the `alpaka_cuda_async` namespace or the `ALPAKA_ACC_GPU_CUDA_ENABLED` macro with HIP/ROCm equivalent code, using the `alpaka_rocm_async` namespace and the `ALPAKA_ACC_GPU_HIP_ENABLED` macro.

4. Update the command line options in `main.cc` and the list of plugins

5. Update the code under the `.../alpaka/...` folders

Mostly, I've changed
```
#ifdef ALPAKA_ACC_GPU_CUDA_ASYNC_BACKEND
```
to
```
#if defined(ALPAKA_ACC_GPU_CUDA_ASYNC_BACKEND) || defined(ALPAKA_ACC_GPU_HIP_ASYNC_BACKEND)
```

6. Unrelated changes

There are also some unrelated changes due to `clang` complaining about implicitly-deleted default constructors, the inappropriate use of `std::move`, and some missing casts.

### Device code changes

The two places with a lot of changes are `src/alpaka/AlpakaCore/prefixScan.h` and `src/alpaka/AlpakaCore/radixSort.h`:

HIP does not support the masked warp instructions like `__shfl_up_sync`, it still has the pre-CUDA 9 versions like `__shfl_up`, so I've `#ifdef`ed them... eventually the whole code should be rewritten using the primitives provided by Alpaka, and benchmarked to make sure that does not introduce any regressions.

I've also added (unconditionally) the memory fence from cms-patatrack/pixeltrack-standalone#210; this too should be benchmarked to check the impact on the CUDA implementation.

### CODE REVIEW

**Comment by fwyzard** [comment]
```
$ ./hip --numberOfThreads 20 --numberOfStreams 20 --maxEvents 2000 --validation
Found 1 devices
Processing 2000 events, of which 20 concurrently, with 20 threads.
CountValidator: all 2000 events passed validation
 Average relative track difference 0.000922907 (all within tolerance)
 Average absolute vertex difference 0.0005 (all within tolerance)
Processed 2000 events in 6.039834e+00 seconds, throughput 331.135 events/s, CPU usage per thread: 31.2%
```

File: Makefile [hunk 1]
```diff
@@ -83,14 +83,16 @@ ifeq ($(wildcard $(ROCM_BASE)),)
...
+export ROCM_DEPS := $(ROCM_LIBDIR)/libamdhip64.so
 export ROCM_HIPCC := $(ROCM_BASE)/bin/hipcc
 HIPCC_UNSUPPORTED_CXXFLAGS := --param vect-max-version-for-alias-checks=50 -Werror=format-contains-nul -Wno-non-template-friend -Werror=return-local-addr -Werror=unused-but-set-variable
 export HIPCC_CXXFLAGS := -fno-gpu-rdc --amdgpu-target=gfx900 $(filter-out $(HIPCC_UNSUPPORTED_CXXFLAGS),$(CXXFLAGS)) --gcc-toolchain=$(GCC_TOOLCHAIN)
 export HIPCC_LDFLAGS := $(LDFLAGS) --gcc-toolchain=$(GCC_TOOLCHAIN)
 # flags to be used by GCC when compiling host code that includes hip_runtime.h
-HIPCONFIG := $(ROCM_BASE)/bin/hipconfig
-export HIP_CXXFLAGS:= $(shell $(HIPCONFIG) --cpp_config)
+export ROCM_CXXFLAGS := -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -I$(ROCM_BASE)/include -I$(ROCM_BASE)/include/hip -I$(ROCM_BASE)/include/hiprand -I$(ROCM_BASE)/include/hsa -I$(ROCM_BASE)/include/rocrand -I$(ROCM_BASE)/llvm/lib/clang/14.0.0
+export ROCM_LDFLAGS := -L$(ROCM_LIBDIR) -L$(ROCM_BASE)/hip/lib -L$(ROCM_BASE)/llvm/lib/clang/14.0.0/lib/linux -L$(ROCM_BASE)/llvm/lib -lamdhip64
```
[the remark this answers is not in the document]
Comment (fwyzard) [reply Makefile:95]: `hipconfig` should be equivalent for the `CXXFLAGS`, but I didn't find a way to make it emit the `LDFLAGS` :-/

[the remark this answers is not in the document]
Comment (fwyzard) [reply Makefile:95]: well, we could try to parse the output of `hipconfig -f`:
```
$ /opt/rocm-5.0.2/bin/hipconfig -f -n
HIP version  : 5.0.13601-ded05588

== hipconfig
HIP_PATH     : /cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hip
ROCM_PATH    : /cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2
HIP_COMPILER : clang
HIP_PLATFORM : amd
HIP_RUNTIME  : rocclr
CPP_CONFIG   :  -D__HIP_PLATFORM_HCC__= -D__HIP_PLATFORM_AMD__= -I/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hip/include -I/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/llvm/bin/../lib/clang/14.0.0 -I/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hsa/include

== hip-clang
HSA_PATH         : /cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hsa
HIP_CLANG_PATH   : /cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/llvm/bin
AMD clang version 14.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.0.2 22065 030a405a181176f1a7749819092f4ef8ea5f0758)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/llvm/bin
AMD LLVM version 14.0.0git
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: znver3

  Registered Targets:
    amdgcn - AMD GCN GPUs
    r600   - AMD GPUs HD2XXX-HD6XXX
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64
hip-clang-cxxflags :  -std=c++11 -isystem "/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/llvm/lib/clang/14.0.0/include/.." -isystem /cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hsa/include -isystem "/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hip/include" -O3
hip-clang-ldflags  :  -L"/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/hip/lib" -O3 -lgcc_s -lgcc -lpthread -lm -lrt

=== Environment Variables
PATH=/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/bin:/cvmfs/cms.cern.ch/rucio/x86_64/cs8/py3/current/bin:/afs/cern.ch/cms/caf/scripts:/cvmfs/cms.cern.ch/common:/cvmfs/grid.cern.ch/emi3ui-latest/bin:/cvmfs/grid.cern.ch/emi3ui-latest/sbin:/cvmfs/grid.cern.ch/emi3ui-latest/usr/bin:/cvmfs/grid.cern.ch/emi3ui-latest/usr/sbin:/opt/rh/gcc-toolset-9/root/usr/bin:/afs/cern.ch/user/f/fwyzard/private/local/bin:/afs/cern.ch/user/f/fwyzard/private/local/bin/x86_64:/afs/cern.ch/user/f/fwyzard/private/local/libexec/x86_64:/afs/cern.ch/user/f/fwyzard/.local/bin:/afs/cern.ch/cms/caf/scripts:/cvmfs/cms.cern.ch/common:/usr/sue/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/puppetlabs/bin
LD_LIBRARY_PATH=/data/user/fwyzard/pixeltrack-standalone/external/tbb/lib:/data/user/fwyzard/pixeltrack-standalone/external/libbacktrace/lib:/data/user/fwyzard/pixeltrack-standalone/external/boost/lib:/data/user/fwyzard/pixeltrack-standalone/external/hwloc/lib:/cvmfs/patatrack.cern.ch/externals/x86_64/rhel8/amd/rocm-5.0.2/lib:/data/user/fwyzard/pixeltrack-standalone/external/cupla/lib:/data/user/fwyzard/pixeltrack-standalone/external/kokkos/install/lib:/cvmfs/grid.cern.ch/emi3ui-latest/lib64:/cvmfs/grid.cern.ch/emi3ui-latest/lib:/cvmfs/grid.cern.ch/emi3ui-latest/usr/lib64:/cvmfs/grid.cern.ch/emi3ui-latest/usr/lib:/opt/rh/gcc-toolset-9/root/usr/lib64:/opt/rh/gcc-toolset-9/root/usr/lib:/opt/rh/gcc-toolset-9/root/usr/lib64/dyninst:/opt/rh/gcc-toolset-9/root/usr/lib/dyninst:/opt/rh/gcc-toolset-9/root/usr/lib64:/opt/rh/gcc-toolset-9/root/usr/lib

== Linux Kernel
Hostname     : patatrack03
Linux patatrack03 5.4.182-1.el8.elrepo.x86_64 #1 SMP Tue Mar 1 13:08:27 EST 2022 x86_64 x86_64 x86_64 GNU/Linux
```

but I have no idea of the variability across different versions.

[the remark this answers is not in the document]
Comment (fwyzard) [reply Makefile:95]: > How much does the `clang/14.0.0` in the paths tie into specific ROCm version?

ROCm 4.5.x uses clang 13.0.0, ROCm 5.0.x uses clang 14.0.0 .

**Comment by fwyzard** [comment]
@makortel this PR has grown to be quite large... let me know if you would rather have it split into smaller ones.

**Comment by fwyzard** [comment]
By the way, I've tested that `kokkostest` builds and run, but I could not get `kokkos` to build, as it would get stuck while compiling or linking some of the tests.

**Comment by makortel** [comment]
> By the way, I've tested that `kokkostest` builds and run, but I could not get `kokkos` to build, as it would get stuck while compiling or linking some of the tests.

The compilation `kokkos` program taking outrageously long for HIP is a known problem, see https://github.com/cms-patatrack/pixeltrack-standalone/issues/178#issuecomment-974859175 and the following discussion (has been reported to Kokkos, link at the bottom of the issue).

**Comment by fwyzard** [comment]
OK, then I won't worry about it.

**Comment by makortel** [comment]
> this PR has grown to be quite large... let me know if you would rather have it split into smaller ones.

Looking at the commits I think splitting this PR into three could be worth it
- Update `Makefile`, `hip', and `hiptest` (first three commits)
- Update Alpaka external and `alpakatest (next two commits)
- Add CUDA xor ROCm/HIP support for `alpaka` and `alpakatest` (last two commits)

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (fwyzard) [reply Makefile]: `clang` itself knows where to find its components, and can ask it about the location, so we can avoid hardcoding it.

**Comment by fwyzard** [comment]
OK, I've split it into
  - #346 [alpaka] Update the develop branch to 2022.03.18 / 343e6de3d62
  - #347 HIP/ROCm updates
  - #348 [alpakatest] Update common files form alpaka
  - #342 (this PR) [alpaka, alpakatest] Support CUDA or ROCm/HIP

#342 (this PR) needs to be merged after #347.

File: src/alpaka/AlpakaCore/radixSort.h [hunk 2]
```diff
@@ -166,8 +172,11 @@ namespace cms::alpakatools {
         });
         alpaka::syncBlockThreads(acc);
 
-        if (threadIdxLocal == 0)
+        if (threadIdxLocal == 0) {
           ibs -= sb;
+          // cms-patatrack/pixeltrack-standalone#210
+          alpaka::mem_fence(acc, alpaka::memory_scope::Grid{});
```
Comment (fwyzard) [inline src/alpaka/AlpakaCore/radixSort.h:178]: We should benchmark the changes on an NVIDIA GPU to see if this has any negative impact on the performance there.

File: src/alpaka/AlpakaCore/alpakaMemory.h [hunk 3]
```diff
@@ -75,12 +78,19 @@ namespace cms::alpakatools {
   template <typename TBuf>
   void pin_buffer(TBuf& buffer) {
 #ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
-    if (not devices<alpaka::PltfUniformCudaHipRt>.empty()) {
-      // it is possible to initialise the CUDA runtime and call cudaHostRegister
-      // only if the system has at least one supported GPU
-      alpaka::prepareForAsyncCopy(buffer);
-    }
+      if (not cms::alpakatools::devices<alpaka::PltfCudaRt>.empty()) {
```
Comment (makortel) [inline src/alpaka/AlpakaCore/alpakaMemory.h:81]: Why the extra indentation? Does it come from `clang-format`?

File: src/alpaka/Makefile [hunk 4]
```diff
@@ -180,6 +211,10 @@ $$($(1)_CUDA_LIB): $$($(1)_CUDA_OBJ) $$($(1)_CUDADLINK) $$(foreach dep,$(EXTERNA
 	@[ -d $$(@D) ] || mkdir -p $$(@D)
 	$(CXX) $$($(1)_CUDA_OBJ) $$($(1)_CUDADLINK) $(LDFLAGS) -shared $(SO_LDFLAGS) $(LIB_LDFLAGS) $$(foreach lib,$$($(1)_DEPENDS),$$($$(lib)_LDFLAGS)) $$(foreach lib,$$($(1)_DEPENDS),$$($$(lib)_CUDA_LDFLAGS)) $$(foreach dep,$(EXTERNAL_DEPENDS),$$($$(dep)_LDFLAGS)) -o $$@
 
+$$($(1)_ROCM_LIB): $$($(1)_ROCM_OBJ) $$(foreach dep,$(EXTERNAL_DEPENDS_H),$$($$(dep)_DEPS)) $$(foreach lib,$$($(1)_DEPENDS),$$($$(lib)_LIB)) $$(foreach lib,$$($(1)_DEPENDS),$$($$(lib)_ROCM_LIB))
+	@[ -d $$(@D) ] || mkdir -p $$(@D)
+	$(CXX) $$($(1)_ROCM_OBJ) $(LDFLAGS) -shared $(SO_LDFLAGS) $(LIB_LDFLAGS) $$(foreach lib,$$($(1)_DEPENDS),$$($$(lib)_LDFLAGS)) $$(foreach lib,$$($(1)_DEPENDS),$$($$(lib)_ROCM_LDFLAGS)) $$(foreach dep,$(EXTERNAL_DEPENDS),$$($$(dep)_LDFLAGS)) -o $$@
```
Comment (makortel) [inline src/alpaka/Makefile:216]: Oh nice, linking object files with ROCm device code works automatically with the host compiler.

File: src/alpaka/AlpakaCore/alpakaMemory.h (hunk 3 above)
Comment (fwyzard) [reply src/alpaka/AlpakaCore/alpakaMemory.h:81]: Mhm, I guess... let me check

Comment (fwyzard) [reply src/alpaka/AlpakaCore/alpakaMemory.h]: Actually, either I didn't run it, or I ran some different version... I've just pushed the changes from `clang-format-12`

File: src/alpaka/Makefile (hunk 4 above)
Comment (fwyzard) [reply src/alpaka/Makefile:216]: On the other hand, with Alpaka (almost) all device code will end up in header files, templated on the accelerator type.
So maybe we can get rid of separable compilation for CUDA as well ?

**Comment by fwyzard** [comment]
Rebased, and squashed the `clang-format` changes.

**Comment by makortel** [comment]
Here is a comparison on V100 with 4 2-minutes jobs
alpaka_cuda_throughput

Within the statistical uncertainty (few %)

**Comment by fwyzard** [comment]
No differences on a T4, either:
image
image

**Comment by fwyzard** [comment]
Allow building the "alpakatest" and "alpaka" applications with support for
either of CUDA or ROCm/HIP.

**Pull request merged**